### PR TITLE
[SYCL] Fix UB and alignment issues in the SYCL default sorter

### PR DESCRIPTION
**Pull request by againull**
Currently `std::byte*` scratch pointer is not aligned and `reinterpret_cast`ed as `T*` where type `T` may have alignment requirement different from `byte*`, this is UB.

As a solution, use `std::align` to align the required buffer in the scratch and use placement `new` so that dynamic type of the buffer in the scratch will be `T*`.

### CODE REVIEW

File: sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp [hunk 1]
```diff
@@ -63,10 +63,25 @@ template <typename Compare = std::less<>> class default_sorter {
...
                   [[maybe_unused]] Ptr last) {
 #ifdef __SYCL_DEVICE_ONLY__
+    // Adjust the scratch pointer based on alignment of the type T.
     // Per extension specification if scratch size is less than the value
-    // returned by memory_required then behavior is undefined, so we don't check
-    // that the scratch size statisfies the requirement.
-    sycl::detail::merge_sort(g, first, last - first, comp, scratch.data());
+    // returned by memory_required, behavior is undefined. So, if std::align
+    // returns nullptr (which means that we don't have enough space) then exit
+    // early.
```
Comment (aelovikov-intel) [inline sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp:70]: I think we can just ignore and continue executing. Seg. fault accessing `nullptr` would be fine.

File: sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp [hunk 2]
```diff
@@ -63,10 +63,25 @@ template <typename Compare = std::less<>> class default_sorter {
...
+    size_t n = last - first;
+    if (!std::align(alignof(T), n * sizeof(T), scratch_ptr, space))
+      return;
+
+    std::size_t local_id = g.get_local_linear_id();
+    T *scratch_begin =
+        local_id == 0 ? scratch_begin = ::new (scratch_ptr) T[n] : nullptr;
+    // Broadcast leader's pointer (the beginning of the scratch) to all work
+    // items in the group.
+    scratch_begin = sycl::group_broadcast(g, scratch_begin);
```
Comment (aelovikov-intel) [inline sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp:83]: If we're broadcasting, should we guard all the code above with `g.is_leader()` as well?

Comment (againull) [reply sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp:81]: Yes, makes sense, fixed.

File: sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp (hunk 1 above)
Comment (againull) [reply sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp]: Sounds good, updated.

File: sycl/include/sycl/detail/group_sort_impl.hpp [hunk 3]
```diff
@@ -71,19 +71,14 @@ struct GetValueType<sycl::multi_ptr<ElementType, Space, IsDecorated>> {
 // since we couldn't assign data to raw memory, it's better to use placement
 // for first assignment
 template <typename Acc, typename T>
-void set_value(Acc ptr, const size_t idx, const T &val, bool is_first) {
-  if (is_first) {
-    ::new (ptr + idx) T(val);
-  } else {
-    ptr[idx] = val;
-  }
+void set_value(Acc ptr, const size_t idx, const T &val) {
```
Comment (aelovikov-intel) [inline sycl/include/sycl/detail/group_sort_impl.hpp:74]: I don't think we need this helper anymore. Sorry for not noticing in the previous review round.

File: sycl/include/sycl/detail/group_sort_impl.hpp [hunk 4]
```diff
@@ -200,10 +189,9 @@ void bubble_sort(Iter first, const size_t begin, const size_t end,
   }
 }
 
-template <typename Group, typename Iter, typename Compare>
-void merge_sort(Group group, Iter first, const size_t n, Compare comp,
-                std::byte *scratch) {
-  using T = typename GetValueType<Iter>::type;
+template <typename Group, typename T, typename Compare>
+void merge_sort(Group group, T *first, const size_t n, Compare comp,
+                T *scratch) {
```
Comment (aelovikov-intel) [inline sycl/include/sycl/detail/group_sort_impl.hpp:194]: Why this change?

Comment (againull) [reply sycl/include/sycl/detail/group_sort_impl.hpp:194]: Now we don't pass the pointer to the scratch as `std::byte *` but do placement new converting it to T*, so we pass it as T* now.

Comment (aelovikov-intel) [reply sycl/include/sycl/detail/group_sort_impl.hpp:194]: I meant `Iter` -> `T*`, not the `std::byte` -> `T` change.

File: sycl/include/sycl/detail/group_sort_impl.hpp (hunk 3 above)
Comment (againull) [reply sycl/include/sycl/detail/group_sort_impl.hpp]: I agree, removed.

File: sycl/include/sycl/detail/group_sort_impl.hpp (hunk 4 above)
Comment (againull) [reply sycl/include/sycl/detail/group_sort_impl.hpp]: Indeed, that's unnecessary limitation, reverted.

**Review by aelovikov-intel — APPROVED** [review]

File: sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp [hunk 5]
```diff
@@ -63,10 +63,33 @@ template <typename Compare = std::less<>> class default_sorter {
...
+    scratch_ptr = std::align(alignof(T), n * sizeof(T), scratch_ptr, space);
+    std::byte *aligned_scratch_ptr = static_cast<std::byte *>(scratch_ptr);
+    const size_t idx = g.get_local_linear_id();
+    const size_t local = g.get_local_range().size();
+    T *local_copy = nullptr;
+    // Each work item constructs several elements based on the input size.
+    // Remember pointer to the beginning of the chunk in local_copy.
+    const size_t chunk = (n - 1) / local + 1;
+    for (int i = chunk; i >= 0; --i) {
+      auto index = idx * chunk + i;
```
Comment (aelovikov-intel) [inline sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp:83]: I think something like this would be clearer:

```
// Initialize in chunks: 
// WI  |   0              |  1                          | ...
// Idxs| 0..chunk_size -1 | chunk_size..2*chunk_size -1 | ...
for (int i = idx*chunk_size; i < (idx+1)*chunk_size -1; ++i)
```

or, if it would be more beneficial to the performance make it interleaved:
```
for (int i = idx; i < n; i+=local_size)
```

Or are you doing something completely different here?

**Comment by againull** [comment]
Tag @andreyfe1

File: sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp [hunk 6]
```diff
@@ -77,16 +100,30 @@ template <typename Compare = std::less<>> class default_sorter {
...
+    const size_t range_size = g.get_local_range().size();
+    const size_t local_id = g.get_local_linear_id();
+    scratch_ptr =
+        std::align(alignof(T), /* output storage and temporary storage */ 2 *
+                                   range_size * sizeof(T),
+                   scratch_ptr, space);
+    std::byte *aligned_scratch_ptr = static_cast<std::byte *>(scratch_ptr);
+    // Each work item processes two locations: in the output buffer and in the
+    // temp buffer.
+    T *local_copy = ::new (aligned_scratch_ptr + local_id * sizeof(T)) T(val);
```
Comment (andreyfe1) [inline sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp:118]: since `local_id * sizeof(T)` used twice (here and 1 line below) it can be useful to create an additional variable. It would also improve readability

File: sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp [hunk 7]
```diff
@@ -77,16 +100,30 @@ template <typename Compare = std::less<>> class default_sorter {
...
+    ::new (aligned_scratch_ptr + range_size * sizeof(T) + local_id * sizeof(T))
+        T(val);
+    // Broadcast leader's pointer (the beginning of the scratch) to all work
+    // items in the group.
+    T *scratch_begin = sycl::group_broadcast(g, local_copy);
```
Comment (andreyfe1) [inline sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp:123]: Sorry for stupid question. Can we replace `group_broadcast` with more lightweight solution?

**Review by andreyfe1 — COMMENTED** [review]
Looks good to me overall, thanks!

File: sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp (hunk 7 above)
Comment (aelovikov-intel) [reply sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp:123]: I think that is actually a very complicated question... https://en.cppreference.com/w/cpp/utility/launder seems to be related but I don't know how it plays with memory model/WIs synchronization rules/etc. In other words, would it be legal to use it, if some non-leader WI does `std::launder` before leader had a chance to perform placement new?

\+ @gmlueck , if he has some thoughts about it.

File: sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp (hunk 5 above)
Comment (againull) [reply sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp:83]: I am iterating backwards so that I save pointer to the first element in the chunk because I need to broadcast pointer to the beginning of the scratch later.

File: sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp (hunk 7 above)
Comment (gmlueck) [reply sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp:123]: I didn't look at the code enough to understand how the algorithm works.  However, the `group_broadcast` operation also acts as a barrier, waiting for all work-items to arrive before any work-item progresses beyond the call to `group_barrier`.  Calling `std::launder` wouldn't have this effect.  Does the algorithm expect a barrier operation here?

**Review by aelovikov-intel — APPROVED** [review]
